feat(ekf2): Model-based 3D wind estimator for fixed-wing

Jaeyoung-Lim opened the pull request:
### Solved Problem
Fixed-wing vehicles are sensitive to wind, but the current wind estimation only considers 2D wind. This is because the vertical component of the wind is unobservable with the pitot tube.


### Solution
This PR implements a 3D wind estimator, which estimates the airflow angles relative to the vehicle. This implementation uses a first principle models with some simplifications. 

- Implementation of a model-based air vector estimator
- Wind estimation based on projected estimated airflow vectors
- Added airlfow.msg to log air relative vectors
- Added groundtruth comparison with simulation for evaluation

While the implemented model is a bit simplistic, I believe this should be a good place to start and add more sophisticated algorithms, with the simulated evaluation environment. 

The implementation comes from the following [paper](https://arxiv.org/abs/2512.06610).
- Harms, Marvin, Jaeyoung Lim, David Rohr, Friedrich Rockenbauer, Nicholas Lawrance, and Roland Siegwart. "Robust Optimization-based Autonomous Dynamic Soaring with a Fixed-Wing UAV." arXiv preprint arXiv:2512.06610 (2025). 

### Changelog Entry
For release notes:
```
Feature: Model-based 3D wind estimation for fixed-wing vehicles
New parameter: 
    FW_W_MASS: Vehicle Mass
    FW_W_AREA: Wing Area
    FW_W_C_B1: Vehicle Aerodynamic coefficient (side slip)
    FW_W_C_A0: Vehicle Aerodynamic coefficient (zero angle of attack)
    FW_W_C_A1: Vehicle Aerodynamic coefficient (angle of attack)
Documentation: Need to document once implemenation is finished
```

### Alternatives
- I have defined a new `airflow.msg` that includes both air-relative velocities and projected wind estimates. We can maybe track this as angle of attack, but given the small values thought using velocities may be better

### Test coverage
Tested with `advanced_plane` in gz sim
<img width="1448" height="262" alt="image" src="https://github.com/user-attachments/assets/558f3a90-2b1d-4b9b-8a5d-45f101c4444c" />
<img width="1452" height="263" alt="image" src="https://github.com/user-attachments/assets/23825534-986a-464a-b0c8-1a2a1f849b20" />
<img width="1439" height="262" alt="image" src="https://github.com/user-attachments/assets/a73109b8-e8c3-4d35-98e0-8642e25d3947" />



### Context
- This implementation is adapted from @MarvinHarms's masters thesis. Preprint: https://arxiv.org/abs/2512.06610
- @msberk FYI

Jaeyoung-Lim committed:
Add module for model-based 3D wind estimation

Add Airflow message

Integrate airflow messages

initialize parameters at a better place

Poll local velocity for wind estimates

F

Fix

Jaeyoung-Lim committed:
Add groundtruth topics for evaluation

Jaeyoung-Lim committed:
Log airflow topics

bresch commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp:
```diff
@@ -0,0 +1,307 @@
+
+	if (_vehicle_acceleration_sub.update(&vehicle_acceleration)) {
+		Dcmf R_ib(_attitude);
+		_acceleration = R_ib * Vector3f(vehicle_acceleration.xyz);
```
```suggestion
		_acceleration = _attitude.rotateVector(vehicle_acceleration.xyz);
```
This is faster when rotating a single vector.

bresch commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp:
```diff
@@ -0,0 +1,307 @@
+	Dcmf R_ib(_attitude);
+	Dcmf R_bi(R_ib.transpose());
+	// compute expected AoA from g-forces:
+	matrix::Vector3f body_force = _mass * R_bi * (_acceleration + _gravity);
```
```suggestion
	matrix::Vector3f body_force = _mass * _attitude.rotateVectorInverse(_acceleration + _gravity);
```
much faster

bresch commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp:
```diff
@@ -0,0 +1,307 @@
+		if (_calibrated_airspeed > _stall_airspeed) {
+			matrix::Vector3f air_velocity_body = compute_wind_estimate();
+
+			Dcmf R_ib(_attitude);
```
as for the other ones, only convert to `Dcmf` when you need to rotate several vectors

bresch commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.hpp:
```diff
@@ -0,0 +1,156 @@
+	matrix::Quatf 				_attitude{};
+	matrix::Vector3f	_acceleration{0.f, 0.f, 0.f};
+	matrix::Vector3f _local_velocity{0.f, 0.f, 0.f};
+	matrix::Vector3f _gravity{0.f, 0.f, 9.81f};
```
You can have this constant in the geo library ([CONSTANTS_ONE_G](https://github.com/PX4/PX4-Autopilot/blob/0cefd74fee03c4ede891daf640bd6a776b23e620/src/lib/geo/geo.h#L54)) and directly use it in the code instead of keeping as a member.

bresch commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp:
```diff
@@ -0,0 +1,307 @@
+
+	// Get Aerodynamic coefficients
+	_wing_area = _param_fw_w_area.get();
+	_C_B1 = _param_fw_w_c_b1.get();
```
you can simply use `_param_fw_w_c_b1.get()` in the equation as this has virtually no overhead. No need to save a copy of it.

Same for all the other parameters.

bresch reviewed (changes requested):
Pretty cool!

Jaeyoung-Lim committed:
Fix logging and comment out printouts

Jaeyoung-Lim committed:
Update src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp

Co-authored-by: Mathieu Bresciani <brescianimathieu@gmail.com>
Update src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp

Co-authored-by: Mathieu Bresciani <brescianimathieu@gmail.com>
Update src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp

Co-authored-by: Mathieu Bresciani <brescianimathieu@gmail.com>
Update src/modules/fw_wind_estimator/FixedwingWindEstimator.hpp

Co-authored-by: Mathieu Bresciani <brescianimathieu@gmail.com>
Update src/modules/fw_wind_estimator/fw_wind_estimator_params.c

Co-authored-by: Mathieu Bresciani <brescianimathieu@gmail.com>
Address review comments

Jaeyoung-Lim commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp:
```diff
@@ -0,0 +1,307 @@
+	// ***************** NEW COMPUTATION FROM MATLAB CALIBRATION **********************
+	float speed = fmaxf(_calibrated_airspeed, _stall_airspeed);
+	float u_approx = _true_airspeed;
+	float v_approx = body_force(1) * _true_airspeed / (0.5f * _rho * powf(speed, 2) * _wing_area * _C_B1);
```
But by using calibrated airspeed, we can use the constant _rho right?

Jaeyoung-Lim committed:
Fix transformation issues and use aero params from advanced plane

Jaeyoung-Lim committed:
Parse parameters without intermediate representations

Jaeyoung-Lim committed:
Use gravity constant

Jaeyoung-Lim committed:
Remove Dcmf

Jaeyoung-Lim committed:
Use new param yaml format

Jaeyoung-Lim committed:
Update units

Jaeyoung-Lim committed:
Remove units

bresch commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp:
```diff
@@ -0,0 +1,300 @@
+	const float C_A1 = _param_fw_w_c_a1.get();
+
+	// compute expected AoA from g-forces:
+	matrix::Vector3f body_force = _mass * (_acceleration + _attitude.rotateVectorInverse(matrix::Vector3f(0.f, 0.f, CONSTANTS_ONE_G)));
```
```suggestion
	Vector3f body_force = _mass * (_acceleration + _attitude.rotateVectorInverse(Vector3f(0.f, 0.f, CONSTANTS_ONE_G)));
```

bresch commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp:
```diff
@@ -0,0 +1,300 @@
+void
+FixedwingWindEstimator::vehicle_attitude_poll()
+{
+	if (_vehicle_attitude_sub.update(&_vehicle_attitude)) {
```
```suggestion
	vehicle_attitude_s vehicle_attitude;
	if (_vehicle_attitude_sub.update(&vehicle_attitude)) {
```
no need to save the whole message. Same for the other messages

bresch commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp:
```diff
@@ -0,0 +1,300 @@
+{
+
+	_mass = _param_fw_w_mass.get();
+	_stall_airspeed = _param_fw_airspd_stall.get();
```
those can also be used in the code directly

msberk commented:
Thanks for the FYI, this is interesting! @jwhgit01 how does this compare to some of the MB wind estimation work you've done with PX4?

Jaeyoung-Lim committed:
Address review comments

Address comments

Fix

Remove bool vtol

F

F

hamishwillee commented:
@Jaeyoung-Lim Very cool! FMI does this replace the old estimator or live in parallel? 

Anyway, what I'm wondering about is docs? What changes for users in order to use this? Are there any risks we need to highlight?

Jaeyoung-Lim commented on msg/Airflow.msg:
```diff
@@ -0,0 +1,15 @@
+# Airspeed data from sensors
```
Yeah, not sure actually. This is a psuedo measurement, where airspeed is guessed from the model. However, probably the value of publishing it to mavlink is to see whether the sensor is working.

hamishwillee replied:
Thanks. I mean - can you update this description so that someone else seeing `Airspeed.msg` knows what different purposes they serve.

Jaeyoung-Lim committed:
Update message description

Jaeyoung-Lim commented:
@hamishwillee Are the message docs manually generated?

bresch commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.hpp:
```diff
@@ -0,0 +1,146 @@
+
+	//Vehicle parameters
+
+
```
```suggestion
```

bresch commented on src/modules/logger/logged_topics.cpp:
```diff
@@ -155,6 +155,7 @@ void LoggedTopics::add_default_topics()
 	add_topic("fixed_wing_longitudinal_setpoint");
 	add_topic("longitudinal_control_configuration");
 	add_topic("lateral_control_configuration");
+	add_topic("airflow");
```
maybe limit to 10Hz?

bresch commented on src/modules/simulation/gz_bridge/GZBridge.cpp:
```diff
@@ -582,6 +582,22 @@ void GZBridge::poseInfoCallback(const gz::msgs::Pose_V &msg)
+			// Body-relateive air velocity
+			matrix::Dcmf Rib(matrix::Quatf{(float)q_nb.W(), (float)q_nb.X(), (float)q_nb.Y(), (float)q_nb.Z()});
+			matrix::Vector3f velocity_f = matrix::Vector3f{(float)velocity(0), (float)velocity(1), (float)velocity(2)};
+			matrix::Vector3f body_velocity = Rib.transpose() * velocity_f;
```
```suggestion
			matrix::Vector3f body_velocity = matrix::Quatf(vehicle_attitude_groundtruth.q).rotateVectorInverse(velocity_f);
```

And can't you just use `matrix::Vector3f(velocity)` to convert from double to float?

bresch reviewed (changes requested):
I think we're almost good to go. The only blurry point is "how are we going to tune that on a new plane"?

Jaeyoung-Lim committed:
Address review comments

Limit logged frequency

Remove white spaces

hamishwillee commented:
> @hamishwillee Are the message docs manually generated?

No. You don't need to update them, but you might get flaws reported.

The reason is that they are supposed to be updated in CI but I am still waiting on @mrpollo to automate running them in CI. What usually happens is that I manually update them once a week. Yes, it is a colossal waste of my time, but not as much as trying to get CI working.

hamishwillee commented:
>  Maybe we can do something like injecting maneuvers as done in autotuning and get a reasonable model on the other end. This would also allow us to utilize more complex aero models for wind estimation (and control) with better confidence.

FYI I believe this is what ardupilot does (on all frames) to refine the model https://ardupilot.org/copter/docs/common-systemid-mode-operation.html

hamishwillee reviewed (dismissed):
My comments addressed thanks - other than "this needs docs" as per https://github.com/PX4/PX4-Autopilot/pull/25966#issuecomment-3584218247

farhangnaderi commented:
@sfuhrer Would appreciate if you could take a look as well.

sfuhrer commented on src/modules/fw_wind_estimator/module.yaml:
```diff
@@ -0,0 +1,51 @@
+module_name: fw_wind_estimator
+parameters:
+- group: FW Wind Estimator
```
```suggestion
- group: 3D Wind Estimator
```
To clearly distinguish it from the existing (2D) wind estimators?

sfuhrer commented on src/modules/simulation/gz_bridge/GZBridge.cpp:
```diff
@@ -582,6 +582,20 @@ void GZBridge::poseInfoCallback(const gz::msgs::Pose_V &msg)
+			airflow_groundtruth.u = body_velocity(0);
+			airflow_groundtruth.v = body_velocity(1);
+			airflow_groundtruth.w = body_velocity(2);
+			///TODO: Include wind
```
What's missing for that?

Jaeyoung-Lim replied:
A wind topic for uORB

sfuhrer reviewed (commented):
Would have been great if we have an easy to follow guide and pipeline for how to get to the coefficients prior to merging it! Is it realistic to get something usable without alpha and beta vanes?

Jaeyoung-Lim committed:
Update src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp

Co-authored-by: Silvan Fuhrer <silvan@auterion.com>

Jaeyoung-Lim committed:
Update src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp

Co-authored-by: Silvan Fuhrer <silvan@auterion.com>

Jaeyoung-Lim committed:
Update src/modules/fw_wind_estimator/module.yaml

Co-authored-by: Silvan Fuhrer <silvan@auterion.com>

Jaeyoung-Lim committed:
Update src/modules/fw_wind_estimator/module.yaml

Co-authored-by: Silvan Fuhrer <silvan@auterion.com>

Jaeyoung-Lim committed:
Add paper reference in the comments

sfuhrer commented on src/modules/fw_wind_estimator/FixedwingWindEstimator.cpp:
```diff
@@ -0,0 +1,292 @@
+	Vector3f body_force = mass * (_acceleration + _attitude.rotateVectorInverse(Vector3f(0.f, 0.f, CONSTANTS_ONE_G)));
+
+	const float speed = fmaxf(_calibrated_airspeed, stall_airspeed);
+	const float dynamic_force = 0.5f * atmosphere::kAirDensitySeaLevelStandardAtmos * powf(speed, 2) * wing_area;
```
Would make sense to me yeah

Jaeyoung-Lim replied:
Fixed

sfuhrer commented:
> Would there be some middle ground where it would be acceptable?

That's what I was hoping for, but I understand if you think that there is no point of providing a 50 paragraph documentation for how to get tot he ID values. But in that case you could still add a new section in https://docs.px4.io/main/en/modules/modules_main, write down the list of needed parameters and mention that there currently doesn't exist a full pipeline to get them. For me that would be sufficient to merge this PR.

BTW, how about generally renaming it to `3d_wind_estimator` instead of `fw_wind_estimator`? Also for the module name etc. To distinguish it from the existing 2d ones.

Jaeyoung-Lim committed:
Update module name

Jaeyoung-Lim committed:
Improve parameter names

Jaeyoung-Lim committed:
Remove vtol argument

Jaeyoung-Lim committed:
Cleanup dead includes

Jaeyoung-Lim committed:
Use measured air density

Use true airspeed

Jaeyoung-Lim commented on src/modules/wind_estimator_3d/WindEstimator3D.hpp:
```diff
@@ -0,0 +1,143 @@
+	hrt_abstime _last_run{0};
+
+	float _calibrated_airspeed{0.0f};
+	float _true_airspeed{15.0f};
```
Fixed

Jaeyoung-Lim commented on src/modules/wind_estimator_3d/module.yaml:
```diff
@@ -0,0 +1,51 @@
+      increment: 0.01
+    FW_W_C_A1:
+      description:
+        short: Vehicle Aerodynamic coefficient (angle of attack)
```
Hehe

Jaeyoung-Lim committed:
Add docs for wind estimator

Jaeyoung-Lim committed:
Add airflow message to ToC

Jaeyoung-Lim commented:
@sfuhrer @bresch @hamishwillee I tried to address the comments in the PR. Would it be possible for you to have another look?

Thanks!

hamishwillee commented on docs/en/modules/modules_estimator.md:
````diff
@@ -113,3 +113,26 @@ mc_hover_thrust_estimator <command> [arguments...]
    status        print status info
 ```
+
+## wind_estimator_3d
````
FYI you don't have to build the docs for modules, parameters, msgs: this gets done for you on merge. That said, I don't think it matters any more ... because this gets done on merge.

Also, someone needs to add them to the sidebar, so the fact you have build and done that is good.

hamishwillee commented on msg/Airflow.msg:
```diff
@@ -0,0 +1,14 @@
+float32 w               	 # [m/s] Body-relative air relative velocity component / Z direction
+float32 windspeed_north	# [m/s] Wind component in north / X direction
+float32 windspeed_east # [m/s] Wind component in east / Y direction
+float32 windspeed_down # [m/s] Wind component in down / Z direction
```
The docs standard now says one space.

```suggestion
uint64 timestamp # [us] Time since system start
uint64 timestamp_sample # [us] Timestamp of the raw data
float32 u # [m/s] Body-relative air relative velocity component / X direction
float32 v # [m/s] Body-relative air relative velocity component / Y direction
float32 w # [m/s] Body-relative air relative velocity component / Z direction
float32 windspeed_north # [m/s] Wind component in north / X direction
float32 windspeed_east # [m/s] Wind component in east / Y direction
float32 windspeed_down # [m/s] Wind component in down / Z direction
```

hamishwillee committed:
docs(docs): msg file whitespace

hamishwillee commented on msg/Airflow.msg:
```diff
@@ -0,0 +1,14 @@
+
+uint64 timestamp # [us] Time since system start
+uint64 timestamp_sample # [us] Timestamp of the raw data
+float32 u # [m/s] Body-relative air relative velocity component / X direction
+float32 v # [m/s] Body-relative air relative velocity component / Y direction
+float32 w # [m/s] Body-relative air relative velocity component / Z direction
```
Just FMI, why u, v, w ?

hamishwillee commented on docs/en/features_fw/wind_estimator_3d.md:
```diff
@@ -0,0 +1,49 @@
+# 3D Wind Estimator
+
+<Badge type="warning" text="Experimental" />
```
Added badge on assumption is that this will go into PX4 v1.19, but we don't know yet.

Can you also add a release note in the main release notes?

Docs look good. I subedited for some of the more common docs conventions and docs should not block merging when this PR is ready.

hamishwillee committed:
docs(docs): subedit

hamishwillee commented on docs/en/features_fw/wind_estimator_3d.md:
```diff
@@ -0,0 +1,65 @@
+# 3D Wind Estimator
+
+<Badge type="tip" text="main (PX4 v1.19)" /> <Badge type="warning" text="Experimental" />
```
Actually, next version confirmed to be v2.0


```suggestion
<Badge type="tip" text="main (PX4 v2.0)" /> <Badge type="warning" text="Experimental" />
```

Jaeyoung-Lim replied:
Wait we have a 2.0?

hamishwillee replied:
@Jaeyoung-Lim We will have. That's why it is "main" first.
